### ISSUE TITLE
Fails to build with GCC 6

### ISSUE THREAD

**Issue by bzed** [vmware/open-vm-tools#67]
Hi,

as reported in http://bugs.debian.org/812046 open-vm-tools fails to build with gcc/g++ 6. The first error is clearly a bug which affects all compilers. Let me know if I can help to fix these bugs. Fixes would be very appreciated, as gcc6 is supposed to become the default compiler for the next Debian release.

```
> sbuild (Debian sbuild) 0.67.0 (26 Dec 2015) on dl580gen9-02.hlinux
...
> UILDDIR>>/lib/include -I/<<PKGBUILDDIR>>/lib/include -Wdate-time -D_FORTIFY_SOURCE=2 -DUSING_AUTOCONF=1 -DOPEN_VM_TOOLS -DUSE_VGAUTH -I/usr/include -DUSE_ICU -DHAVE_ICU_38 -DVMX86_TOOLS -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -D_XOPEN_SOURCE=500 -D_BSD_SOURCE -D_SVID_SOURCE -D_DEFAULT_SOURCE -DUSE_SSL_DIRECT -fPIC -Wno-deprecated-declarations -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wall -Werror -Wno-pointer-sign -Wno-unused-value -fno-strict-aliasing -Wno-unknown-pragmas -Wno-uninitialized -Wno-deprecated-declarations -Wno-unused-but-set-variable -c asyncsocket.c  -fPIC -DPIC -o .libs/asyncsocket.o
> In file included from /<<PKGBUILDDIR>>/lib/include/hostinfo.h:35:0,
>                  from asyncsocket.c:58:
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:912:51: error: result of '-2147483648 << 1' requires 33 bits to represent, but 'int' only has 32 bits [-Werror=shift-overflow=]
>  #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
>                                                    ^
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:916:34: note: in expansion of macro 'VMW_BIT_MASK'
>     CPUID_##name##_MASK         = VMW_BIT_MASK(size) << bitpos, \
>                                   ^~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:292:1: note: in expansion of macro 'FIELD'
>  FIELD(  0,  0, EAX,  0, 32, NUMLEVELS,                             ANY, FALSE) \
>  ^~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:870:4: note: in expansion of macro 'CPUID_FIELD_DATA_LEVEL_0'
>     CPUID_FIELD_DATA_LEVEL_0                                           \
>     ^~~~~~~~~~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:927:4: note: in expansion of macro 'CPUID_FIELD_DATA'
>     CPUID_FIELD_DATA
>     ^~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:912:51: error: result of '-2147483648 << 1' requires 33 bits to represent, but 'int' only has 32 bits [-Werror=shift-overflow=]
>  #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
>                                                    ^
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:918:34: note: in expansion of macro 'VMW_BIT_MASK'
>     CPUID_INTERNAL_MASK_##name  = VMW_BIT_MASK(size) << bitpos, \
>                                   ^~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:292:1: note: in expansion of macro 'FIELD'
>  FIELD(  0,  0, EAX,  0, 32, NUMLEVELS,                             ANY, FALSE) \
>  ^~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:870:4: note: in expansion of macro 'CPUID_FIELD_DATA_LEVEL_0'
>     CPUID_FIELD_DATA_LEVEL_0                                           \
>     ^~~~~~~~~~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:927:4: note: in expansion of macro 'CPUID_FIELD_DATA'
>     CPUID_FIELD_DATA
>     ^~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:912:51: error: result of '-2147483648 << 1' requires 33 bits to represent, but 'int' only has 32 bits [-Werror=shift-overflow=]
>  #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
>                                                    ^
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:916:34: note: in expansion of macro 'VMW_BIT_MASK'
>     CPUID_##name##_MASK         = VMW_BIT_MASK(size) << bitpos, \
>                                   ^~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:293:1: note: in expansion of macro 'FIELD'
>  FIELD(  0,  0, EBX,  0, 32, VENDOR1,                               YES, TRUE)  \
>  ^~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:870:4: note: in expansion of macro 'CPUID_FIELD_DATA_LEVEL_0'
>     CPUID_FIELD_DATA_LEVEL_0                                           \
>     ^~~~~~~~~~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:927:4: note: in expansion of macro 'CPUID_FIELD_DATA'
>     CPUID_FIELD_DATA
>     ^~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:912:51: error: result of '-2147483648 << 1' requires 33 bits to represent, but 'int' only has 32 bits [-Werror=shift-overflow=]
>  #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
>                                                    ^
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:918:34: note: in expansion of macro 'VMW_BIT_MASK'
>     CPUID_INTERNAL_MASK_##name  = VMW_BIT_MASK(size) << bitpos, \
>                                   ^~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:293:1: note: in expansion of macro 'FIELD'
>  FIELD(  0,  0, EBX,  0, 32, VENDOR1,                               YES, TRUE)  \
>  ^~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:870:4: note: in expansion of macro 'CPUID_FIELD_DATA_LEVEL_0'
>     CPUID_FIELD_DATA_LEVEL_0                                           \
>     ^~~~~~~~~~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:927:4: note: in expansion of macro 'CPUID_FIELD_DATA'
>     CPUID_FIELD_DATA
>     ^~~~~~~~~~~~~~~~
> 
> /<<PKGBUILDDIR>>/lib/include/x86cpuid.h:912:51: error: result of '-2147483648 << 1' requires 33 bits to represent, but 'int' only has 32 bits [-Werror=shift-overflow=]
>  #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
```

**Comment by Conan-Kudo** [comment]
This has also shown up in Fedora's GCC 6 mass rebuild. See [RhBug 1305108](https://bugzilla.redhat.com/show_bug.cgi?id=1305108).

**Comment by AdamWill** [comment]
You can get past that by dropping `-Werror` from CFLAGS and/or changing the `VMW_BIT_MASK` definition, but then you hit a subsequent problem, a bunch of these:

    In file included from /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/hostinfo.h:35:0,
                     from dndGuest/rpcV3Util.cpp:45:
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:912:51: warning: result of '(-2147483648 << 1)' requires 33 bits to represent, but 'int' only has 32 bits [-Wshift-overflow=]
     #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)
                                    ~~~~~~~~~~~~~~~~~~~^~~~
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:916:34: note: in expansion of macro 'VMW_BIT_MASK'
        CPUID_##name##_MASK         = VMW_BIT_MASK(size) << bitpos, \
                                      ^~~~~~~~~~~~
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:292:1: note: in expansion of macro 'FIELD'
     FIELD(  0,  0, EAX,  0, 32, NUMLEVELS,                             ANY, FALSE) \
     ^~~~~
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:870:4: note: in expansion of macro 'CPUID_FIELD_DATA_LEVEL_0'
        CPUID_FIELD_DATA_LEVEL_0                                           \
        ^~~~~~~~~~~~~~~~~~~~~~~~
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:927:4: note: in expansion of macro 'CPUID_FIELD_DATA'
        CPUID_FIELD_DATA
        ^~~~~~~~~~~~~~~~
    /builddir/build/BUILD/open-vm-tools-10.0.0-3000743/lib/include/x86cpuid.h:912:51: error: left operand of shift expression '(-2147483648 << 1)' is negative [-fpermissive]
     #define VMW_BIT_MASK(shift)  (((1 << (shift - 1)) << 1) - 1)

**Comment by bzed** [comment]
You don't want to remove -Werror. You want build failures in case of issues. And then fix those issues.

**Comment by AdamWill** [comment]
Yes, we do. But I'm not gonna argue about that here, it's really not relevant.

**Comment by hwti** [comment]
Using an unsigned constant fixes the errors and warnings :
#define VMW_BIT_MASK(shift)  (((1**u** << (shift - 1)) << 1) - 1)

**Comment by scaronni** [comment]
Yep, compilation goes on without errors or warnings, and then stops with this:

```
linuxDeployment.c:84:18: error: 'CUST_DNS_ERROR' defined but not used [-Werror=unused-const-variable=]
 static const int CUST_DNS_ERROR     = 252;
                  ^~~~~~~~~~~~~~
linuxDeployment.c:81:18: error: 'CUST_GENERIC_ERROR' defined but not used [-Werror=unused-const-variable=]
 static const int CUST_GENERIC_ERROR = 255;
                  ^~~~~~~~~~~~~~~~~~
linuxDeployment.c:80:18: error: 'CUST_SUCCESS' defined but not used [-Werror=unused-const-variable=]
 static const int CUST_SUCCESS       = 0;
                  ^~~~~~~~~~~~
linuxDeployment.c:70:20: error: 'ENDOFLINEMARKER' defined but not used [-Werror=unused-const-variable=]
 static const char  ENDOFLINEMARKER = '\n';
                    ^~~~~~~~~~~~~~~
```

**Comment by scaronni** [comment]
Can we just delete those?

**Comment by pabs3** [comment]
I created the attached patch and tested that it fixes this issue.

**Comment by pabs3** [comment]
Properly attach this time.

**Comment by pabs3** [comment]
Finally figured out how to attach files. github does not support attaching .patch files and does not give errors for that unless JavaScript is on.

[0001-Fix-build-failure-with-GCC-6.patch.txt](https://github.com/vmware/open-vm-tools/files/367722/0001-Fix-build-failure-with-GCC-6.patch.txt)

**Comment by bzed** [comment]
Please nmu if you are in a hurry.

Thanks, 
Bernd 

Am 17. Juli 2016 10:14:11 MESZ, schrieb Paul Wise <notifications@github.com>:
>I created the attached patch and tested that it fixes this issue. 
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/vmware/open-vm-tools/issues/67#issuecomment-233170805

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by pabs3** [comment]
As long as this gets fixed in time for open-vm-tools to migrate back into stretch (it got removed), I'll be happy, so no particular rush but the earlier the better.

**Comment by bzed** [comment]
@oliverkurth @ravindravmw could you guys please review the patch we have in Debian now - or come up with your own fix? Thanks a lot!

**Comment by ravindravmw** [comment]
@bzed we have already applied our patches for gcc6 compilation issues in Fedora packages here, http://pkgs.fedoraproject.org/cgit/rpms/open-vm-tools.git/tree/vmw-bitmask-gcc6.patch and http://pkgs.fedoraproject.org/cgit/rpms/open-vm-tools.git/tree/no-unused-const.patch. Please consider taking these patches because these will be included in the next release of open-vm-tools too.

**Comment by pabs3** [comment]
On Mon, 2016-07-25 at 11:44 -0700, Ravindra Kumar wrote:

> > http://pkgs.fedoraproject.org/cgit/rpms/open-vm-tools.git/tree/no-unused-const.patch.

This one looks incorrect compared to the Debian changes for fixing the same issue.

-- 
bye,
pabs

http://bonedaddy.net/pabs3/

**Comment by bzed** [comment]
@ravindravmw @pabs3 is right here, your workaround with -Wno-unused-const-variable is not really what we prefer to have in Debian.

**Comment by ravindravmw** [comment]
@bzed the workaround I have provided is to get the build go through without you getting blocked. We will do the code cleanup in a subsequent open-vm-tools release. Having said that, any code changes you make on top of what we are publishing as part of open-vm-tools will make it hard for us to support these different open-vm-tools builds. Could you please create a pull request for the Debian patch so that it can be included in a future open-vm-tools release?

**Comment by bzed** [comment]
@ravindravmw @oliverkurth well, my patch got a bit rusty already and is still not merged as it seems. Although it might be hard for you to support different open-vm-tools builds, you might realize that we prefer proper code quality in Debian. And looking at other distributions they seem to think the same way. You might want to merge and fix this properly.

**Comment by ravindravmw** [comment]
@bzed thanks for the pull request. I'm sorry, we could not take your fix for open-vm-tools/lib/include/x86cpuid.h because internally we fixed it differently. You can see the latest code on github and this should work fine on newer gcc - https://github.com/vmware/open-vm-tools/blob/stable-10.1.x/open-vm-tools/lib/include/x86cpuid.h#L978. For your fixes in open-vm-tools/libDeployPkg/linuxDeployment.c, we can start the process for picking the fixes once you sign our CLA. I don't see any update from our CLABot on your pull request. Looks like your pull request was submitted before we enabled our CLABot. You could either create a new pull request to trigger an update from our CLABot or sign our CLA online at https://cla.vmware.com/ to move forward with this pull request. Sorry for the inconvenience. Thanks!

**Comment by bzed** [comment]
@ravindravmw I've signed the CLA one year ago - at least thats what the CLA page says. Should have been signed before this PR.
As the master branch is rather outdated - do you want a PR against stable-10.1.10???

**Comment by ravindravmw** [comment]
Ok thanks. I will check in our system and take it from there. There is no need to submit another pull request at this time.

**Comment by ravindravmw** [comment]
FWIW, master branch has been updated now.

**Comment by ravindravmw** [comment]
@bzed I have confirmed that your CLA is in place. I have raised an VMware internal bug to include your contribution in a future release. Please note that there may be additional modifications to your patch based on other ongoing changes in the code and VMware internal review process.

**Comment by stanguturi** [comment]
Fixed in commit 1604e7c59f9491e0f98e39dc511a534ffea68879 on devel branch.
